p2panda-rs/src/storage_provider/traits/storage_provider.rs

```diff
@@ -0,0 +1,512 @@
+                let entry_hash_backlink = entry_backlink.hash();
+                // Determine skiplink ("lipmaa"-link) entry in this log
+                let entry_hash_skiplink = self.determine_skiplink(&entry_latest).await?;
+
+                Ok(Self::EntryArgsResponse::new(
+                    Some(entry_hash_backlink.clone()),
+                    entry_hash_skiplink,
+                    entry_latest.seq_num().clone().next().unwrap(),
+                    entry_latest.log_id(),
+                ))
+            }
+            // No entry was given yet, we can assume this is the beginning of the log
+            None => Ok(Self::EntryArgsResponse::new(
+                None,
+                None,
+                SeqNum::default(),
+                log,
+            )),
+        }
+    }
+
+    /// Stores an author's Bamboo entry with operation payload in database after validating it.
+    async fn publish_entry(
+        &self,
+        params: &Self::PublishEntryRequest,
+    ) -> Result<Self::PublishEntryResponse, Box<dyn std::error::Error>> {
+        // Create a storage entry.
+        let entry = StorageEntry::new(params.entry_signed(), params.operation_encoded())?;
+        // Validate the entry (this also maybe happened in the above constructor)
+        entry.validate().map_err(|error| format!("{:?}", error))?;
+
+        // Every operation refers to a document we need to determine. A document is identified by the
+        // hash of its first `CREATE` operation, it is the root operation of every document graph
+        let document_id = if entry.operation().is_create() {
+            // This is easy: We just use the entry hash directly to determine the document id
+            DocumentId::new(entry.hash().into())
+        } else {
+            // For any other operations which followed after creation we need to either walk the operation
+            // graph back to its `CREATE` operation or more easily look up the database since we keep track
+            // of all log ids and documents there.
+            //
+            // We can determine the used document hash by looking at what we know about the previous
+            // entry in this author's log.
+            //
+            // @TODO: This currently looks at the backlink, in the future we want to use
+            // "previousOperation", since in a multi-writer setting there might be no backlink for
+            // update operations! See: https://github.com/p2panda/aquadoggo/issues/49
+            let backlink_entry_hash = entry
+                .backlink_hash()
+                .ok_or_else(|| PublishEntryError::OperationWithoutBacklink(entry.hash()))?;
+
+            self.get_document_by_entry(&backlink_entry_hash)
+                .await?
+                // @TODO this trips if the backlink is missing not necessarily the document
+                // needs revisiting when we use previous_operations to retrieve the correct
+                // document.
```

Comment:
??

Reply:
Its a little bit hard to read, I think the problem might be the missing punctuation between the lines :joy_cat: 

How about?

> This error seems to occur when the backlink is missing instead of the document as stated. It should be revisisted when we use previous_operations to retrieve the correct document.

Reply:
Ah! So punctuation _is_ important after all 🤣 who knew?!